### PR TITLE
Add a safe_store feature for token attributes

### LINKED ISSUE #2794: Tokeninfo has been deleted or its row is not present
In a Galera Cluster (or maybe similar multi node setups) the creation of a token can fail with

    ObjectDeletedError: Instance '<TokenInfo at 0x7fcef635aa50>' has been deleted, or its row is otherwise not present.

The error occurs at these code lines in models.py:

            db.session.add(self)
            db.session.commit()
            ret = self.id

This sometimes also happens with the tokenowner, but mostly with tokeninfo.

This seems to be caused by that the creation of the tokeninfo is not synced to all nodes and the node, where the tokeninfo was written in the first place has no information on the DB id, yet.

Field tests indicate, that we can change the code this way, to mitigate the problem. If the ID really needs to be returned, we can 
Issue a new search for the ID - we should make this configurable:

~~~~python
            db.session.add(self)
            db.session.commit()
            if SAFE_STORE:
                ti = TokenInfo.query.filter_by(token_id=self.token_id,
                                               Key=self.Key).first()
                ret = ti.id
            else:
                ret = self.id
~~~~

To me it looks as the same problem like #2396

### PR DESCRIPTION
**Pull request by cornelinux**
tokeninfos and tokenowner information can fail to
be saved (or fail to return the DB id) in certain
database scenarios. This adds the possibility for a safe store
by reading the ID in a new request.

Closes #2794

### CODE REVIEW

**Review by plettich — CHANGES_REQUESTED** [review]
Looks good so far.
In the future we should make use of the generic `MethodsMixin` and the provided `safe`/`delete` functions more often.

**Comment by cornelinux** [comment]
> Looks good so far.
> In the future we should make use of the generic `MethodsMixin` and the provided `safe`/`delete` functions more often.

I think in many cases the save functions are currently rather specific with some logical checks so these can currently not be mixed in. But I am always in for using less lines of code! :-)

File: doc/installation/system/inifile.rst [hunk 1]
```diff
@@ -55,10 +55,23 @@ of an administrator.
...
 ``PI_INIT_CHECK_HOOK`` is a function in an external module, that will be
 called as decorator to ``token/init`` and ``token/assign``. This function
-takes the ``request`` and ``action`` (either "init" or "assing") as an
+takes the ``request`` and ``action`` (either "init" or "assign") as
 arguments and can modify the request or raise an exception to avoid the
 request being handled.
 
+If you set ``PI_DB_SAFE_STORE`` to *True* the database layer will in the cases
+of ``tokenowner``, ``tokeinfo`` and ``tokenrealm`` read the id of the newly created
+database object in an additional request and not return it directly.
```
[the remark this answers is not in the document]
Comment (cornelinux) [reply doc/installation/system/inifile.rst:64]: Sounds sensible. Will do. Maybe "SELECT"...

File: doc/installation/system/inifile.rst [hunk 2]
```diff
@@ -55,10 +55,23 @@ of an administrator.
...
+
+.. Note:: In certain cases (e.g. with Galera Cluster) it can happen that the database
+   node has no information about the object id directly during the write-process.
+   The database might respond with an error like "object has been deleted or its
+   row is otherwise not present". In this case setting ``PI_DB_SAFE_STORE`  to *True*
```
[the remark this answers is not in the document]
Comment (cornelinux) [reply doc/installation/system/inifile.rst:69]: Will do.

File: doc/installation/system/inifile.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (cornelinux) [reply doc/installation/system/inifile.rst]: Done in 0cada8d6

File: doc/installation/system/inifile.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (cornelinux) [reply doc/installation/system/inifile.rst]: Done in 0cada8d6

File: doc/installation/system/inifile.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (cornelinux) [reply doc/installation/system/inifile.rst]: I also added some more text. I hope you will like it :-)

**Review by plettich — APPROVED** [review]
looks good

**Pull request merged**